README.md
# SRAPI

SRAPI is a Go-based API for the Sports Reference network of websites.

_Current Version: 0.2.3 (Pre-Release)_

| Master | Develop |
|:-:|:-:|
|[![Master](https://travis-ci.com/mtslzr/srapi.svg?branch=master)](https://travis-ci.com/mtslzr/srapi)|[![Develop](https://travis-ci.com/mtslzr/srapi.svg?branch=develop)](https://travis-ci.com/mtslzr/srapi)|

## Installation

### Clone Repo

```bash
git clone git@github.com:mtslzr/srapi
cd srapi
```

### Dependencies

_NOTE: Database filename is currently hard-coded. This will change in the future._

```bash
cat srapi.sql | sqlite3 srapi.db
```

### Build

```bash
go build -v --mod=vendor
```

## Usage

### Run

```bash
go run -v ./...
Starting server on localhost:5000...
```

### Executable

```bash
./srapi
Starting server on localhost:5000...
```

### Endpoints

#### Get Standings

Returns current standings. Requires two-letter ID for sport (e.g. `bs` for baseball).

```
GET /{sport}/standings
```

#### Get Teams

Returns all current teams. Requires two-letter ID for sport (e.g. `bs` for baseball).

```
GET /{sport}/teams
```

#### Get Years

Returns all available years. Requires two-letter ID for sport (e.g. `bs` for baseball).

```
GET /{sport}/years
```

## Contribution

### Testing

```bash
go test -v --mod=vendor ./...
```

### Deployment

_Coming soon..._

CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Added CHANGELOG.
- Added CONTRIBUTERS.
- Added gitignore.
- Added MIT license.
- Added README with basic information.
- Added srapi.go (with dummy function).

srapi_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
)

// Start server and query and endpoint.
func startEndpoint(ep string) (res *httptest.ResponseRecorder) {
	req, _ := http.NewRequest("GET", ep, nil)
	res = httptest.NewRecorder()
	startServer().ServeHTTP(res, req)
	return
}

// Test GetStandings for current baseball standings.
func TestGetStandings(t *testing.T) {
	res := startEndpoint("/bs/standings")
	assert.Equal(t, 200, res.Code, "Expected OK response")
	assert.Contains(t, res.Body.String(), "AL",
		"Expected result contains 'AL'")
	assert.Contains(t, res.Body.String(), "NL",
		"Expected result contains 'NL'")
	assert.Contains(t, res.Body.String(), "West",
		"Expected result contains 'West'")
	assert.Contains(t, res.Body.String(), "Miami Marlins",
		"Expected result contains 'Miami Marlins'")
	assert.Contains(t, res.Body.String(), "Minnesota Twins",
		"Expected result contains 'Minnesota Twins'")
}

// Test GetTeams for current baseball teams.
func TestGetTeams(t *testing.T) {
	res := startEndpoint("/bs/teams")
	assert.Equal(t, 200, res.Code, "Expected OK response")
	assert.Contains(t, res.Body.String(), "Miami Marlins",
		"Expected result contains 'Miami Marlins'")
	assert.Contains(t, res.Body.String(), "Minnesota Twins",
		"Expected result contains 'Minnesota Twins'")
}

// Test GetTeams for current baseball teams.
func TestGetYears(t *testing.T) {
	res := startEndpoint("/bs/years")
	assert.Equal(t, 200, res.Code, "Expected OK response")
	assert.Contains(t, res.Body.String(), "2019",
		"Expected result contains '2019'")
	assert.Contains(t, res.Body.String(), "1876",
		"Expected result contains '1876'")
}

// Test an invalid endpoint for failure
func TestInvalidEndpoint(t *testing.T) {
	res := startEndpoint("/fakeUrl")
	assert.Equal(t, 404, res.Code, "Expected 404 response")
	assert.Contains(t, res.Body.String(), "page not found",
		"Expected proper error message")
	res = startEndpoint("/fakeUrl/partTwo")
	assert.Equal(t, 404, res.Code, "Expected 404 response")
}

// Test an invalid Sport ID for failure
func TestInvalidSport(t *testing.T) {
	res := startEndpoint("/xx/standings")
	assert.Equal(t, 500, res.Code, "Expected server error response")
	assert.Contains(t, res.Body.String(), "sql: no rows in result set",
		"Expected proper error message")
}


standings.go
package main

import (
	"strconv"
	"strings"
	"time"

	"github.com/anaskhan96/soup"
)

// Get Current Standings (Baseball)
func bsStandings(bs Sport) (map[string][]string, error) {
	stands := make(map[string][]string)
	leagues := []string{"AL", "NL"}
	divisions := []string{"E", "C", "W"}
	year := strconv.Itoa(time.Now().Year())

	res, err := soup.Get("https://" + bs.Host + "/" +
		strings.Replace(bs.Standings, "{year}", year, 1))
	if err != nil {
		return stands, err
	}

	doc := soup.HTMLParse(res)
	for _, div := range divisions {
		tables := doc.FindAll("div", "id", "all_standings_"+div)
		for x, table := range tables {
			name := table.Find("h2").Text()
			divName := leagues[x] + " " + strings.Replace(name, " Division", "", 1)
			teams := table.FindAll("a")
			for _, team := range teams {
				stands[divName] = append(stands[divName], team.Text())
			}
		}
	}
	return stands, nil
}


srapi.go
package main

import (
	"database/sql"
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
	_ "github.com/mattn/go-sqlite3"
)

var w http.ResponseWriter

// Sport is model for Database data
type Sport struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Host      string `json:"host"`
	Standings string `json:"standings"`
	Teams     string `json:"teams"`
	Years     string `json:"years"`
}

// Start API and set routes
func startServer() http.Handler {
	srv := mux.NewRouter()
	srv.HandleFunc("/{sport}/standings", getStandings).Methods("GET")
	srv.HandleFunc("/{sport}/teams", getTeams).Methods("GET")
	srv.HandleFunc("/{sport}/years", getYears).Methods("GET")
	srv.Use(rwMiddleware)
	return srv
}

// Get Sport by ID
func getSport(id string) (sport Sport, err error) {
	sport, err = queryDb(id)
	return
}

// Get Current Standings
func getStandings(rw http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	sport, err := getSport(params["sport"])
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	stand, err := bsStandings(sport)
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	out, _ := json.Marshal(stand)
	sendResponse(200, out)
}

// Get All Teams
func getTeams(rw http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	sport, err := getSport(params["sport"])
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	teams, err := bsTeams(sport)
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	out, _ := json.Marshal(teams)
	sendResponse(200, out)
}

// Get All Years
func getYears(rw http.ResponseWriter, r *http.Request) {
	params := mux.Vars(r)

	sport, err := getSport(params["sport"])
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	years, err := bsYears(sport)
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
		return
	}

	out, _ := json.Marshal(years)
	sendResponse(200, out)
}

// Query Database and return Sport
func queryDb(sport string) (row Sport, err error) {
	db, err := sql.Open("sqlite3", "./srapi.db")
	if err != nil {
		return
	}

	err = db.QueryRow("SELECT * FROM sports WHERE id = ?", sport).
		Scan(&row.ID, &row.Name, &row.Host, &row.Standings, &row.Teams, &row.Years)
	return
}

// Export http.ResponseWriter for use in sendResponse()
func rwMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
		w = rw
		next.ServeHTTP(rw, r)
	})
}

// Marshal and send response with chosen statusCode
func sendResponse(code int, js []byte) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	_, err := w.Write(js)
	if err != nil {
		out, _ := json.Marshal(err.Error())
		sendResponse(500, out)
	}
}
